numhask/src/NumHask/Algebra/Abstract/Group.hs

```diff
@@ -0,0 +1,121 @@
+{-# OPTIONS_GHC -Wall #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE ConstrainedClassMethods #-}
+{-# LANGUAGE RoleAnnotations #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE MonoLocalBinds #-}
+-- | The Group hirarchy
+module NumHask.Algebra.Abstract.Group
+      ( Magma(..)
+      , Unital(..)
+      , Semigroup
+      , (<>)
+      , Commutative
+      , Absorbing(..)
+      , Invertible(..)
+      , Idempotent
+      , Monoid
+      , mempty
+      , Group
+      , groupSwap
+      , AbelianGroup
+      )
+      where
+
+import qualified Prelude                       as P
+
+-- * Magma structure
+-- | A <https://en.wikipedia.org/wiki/Magma_(algebra) Magma> is a tuple (T,comb) consisting of
+--
+-- - a type a, and
+--
+-- - a function (comb) :: T -> T -> T
+--
+-- The mathematical laws for a magma are:
+--
+-- - comb is defined for all possible pairs of type T, and
+--
+-- - comb is closed in the set of all possible values of type T
+--
+-- or, more tersly,
+--
+-- > ∀ a, b ∈ T: a comb b ∈ T
+--
+-- These laws are true by construction in haskell: the type signature of 'magma' and the above mathematical laws are synonyms.
+--
+--
+class Magma a where
+  comb :: a -> a -> a
+
+-- | A Unital Magma
+--
+-- > unit comb a = a
+-- > a comb unit = a
+--
+class Magma a =>
+      Unital a where
+  unit :: a
+
+-- | A semigroup is an associative Magma
+--
+-- > (a comb b) comb c = a comb (b comb c)
+class Magma a =>
```

Comment:
I don't agree with redefining Semigroup. We should really be following the rest of the ecosystem and : import `semigroups` if GHC < 8, import Semigroup from base otherwise.

Reply:
I understand the usual Semigroup doesn't have the Magma constraint. This is a fork in the path for the library. Packages that will use numhask will be incompatible with those founded on `base`